File: mesmaeker-mockito/src/main/java/fr/xgouchet/mesmaeker/mockito/MockitoOngoingThrowingClause.java
package fr.xgouchet.mesmaeker.mockito;

import org.mockito.Mockito;

import java.util.function.Consumer;

/**
 * @author Xavier F. Gouchet
 */
public class MockitoOngoingThrowingClause<T> implements OngoingThrowingClause<T> {

    private final T subject;
    private final Throwable[] expectedThrowables;


    public MockitoOngoingThrowingClause(T subject, Throwable[] toBeThrown) {
        this.subject = subject;
        expectedThrowables = toBeThrown;
    }

    @SuppressWarnings("unchecked")
    public MockitoOngoingThrowingClause(T subject, Class<? extends Throwable>... toBeThrown) {
        this.subject = subject;
        expectedThrowables = new Throwable[toBeThrown.length];

        for (int i = 0; i < toBeThrown.length; i++) {
            try {
                expectedThrowables[i] = toBeThrown[i].newInstance();
            } catch (InstantiationException | IllegalAccessException e) {
                throw new RuntimeException(e);
            }
        }
    }

    @Override public OngoingThrowingClause<T> when(Consumer<T> consumer) {
        T stubber = Mockito.doThrow(expectedThrowables).when(subject);
        consumer.accept(stubber);
        return this;
    }
}

File: mesmaeker-mockito/src/main/java/fr/xgouchet/mesmaeker/mockito/MockitoOngoingThrowingClause.kt
package fr.xgouchet.mesmaeker.mockito

import org.mockito.Mockito

/**
 * @author Xavier F. Gouchet
 */
class MockitoOngoingThrowingClause<T> : OngoingThrowingClause<T> {

    private val subject: T
    private val expectedThrowables: Array<Throwable>


    constructor(subject: T, toBeThrown: Array<Throwable>) {
        this.subject = subject
        expectedThrowables = toBeThrown
    }

    constructor(subject: T, vararg toBeThrown: Class<out Throwable>) {
        this.subject = subject
        expectedThrowables = listOf(*toBeThrown)
                .map { it.newInstance() }
                .toTypedArray()
    }

    override fun `when`(block: (T) -> Unit): OngoingThrowingClause<T> {
        return whenever(block)
    }

    override fun whenever(block: (T) -> Unit): OngoingThrowingClause<T> {
        val stubber = Mockito.doThrow(*expectedThrowables).`when`(subject)
        block(stubber)
        return this
    }
}
